docs/contributors.rst

```diff
@@ -42,30 +37,27 @@ Specifically, see all :code:`RUN apt-get install` lines.
 Build and Install
 -----------------
 
-First, uninstall any existing version of Horovod.  Be sure to do this *outside* the Horovod root directory:
+From *inside* the Horovod root directory, install Horovod in develop/editable mode:
 
 .. code-block:: bash
 
-    $ cd $HOME
-    $ pip uninstall -y horovod
-    $ cd -
-
-From *inside* the Horovod root directory, remove any previous build artifacts and then install Horovod:
-
-.. code-block:: bash
-
-    $ rm -rf build/ dist/
-    $ HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_TENSORFLOW=1 python setup.py install
+    $ HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_TENSORFLOW=1 pip install -v -e .
 
 Set ``HOROVOD_WITHOUT_[FRAMEWORK]=1`` to disable building Horovod plugins for that framework.
 This is useful when you’re testing a feature of one framework in particular and wish to save time.
+Set ``HOROVOD_WITH_[FRAMEWORK]=1`` to generate an error if the Horovod plugin for that framework failed to build.
+Set ``HOROVOD_GPU_OPERATIONS=NCCL`` to run on GPUs with NCCL.
+You can install optional dependencies defined in `setup.py <https://github.com/horovod/horovod/blob/master/setup.py>`__ by adding brackets
+at the end of the command line e.g. ``[test]`` for test dependencies.
 
-For a debug build with checked assertions etc. replace the invocation of setup.py by:
+In develop mode, you can edit the Horovod source directly in the repo folder. For Python code, the changes will take effect
+immediately. For **C++/CUDA code**, the ``... pip install -v -e .`` command needs to be invoked again to perform an incremental build.
 
-.. code-block:: bash
+For a debug build with debug symbols, checked assertions etc., use the extra flags below:
 
-    $ HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_TENSORFLOW=1 python setup.py build_ext --debug install
+.. code-block:: bash
 
+    $ HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_TENSORFLOW=1 pip install --global-option build_ext --global-option --debug -v -e .
```

Comment:
I just tried to do an editable debug build with a variation of that command line:

```
HOROVOD_WITH_MPI=1 HOROVOD_WITH_PYTORCH=1 HOROVOD_WITHOUT_MXNET=1 HOROVOD_WITHOUT_TENSORFLOW=1 HOROVOD_GPU_OPERATIONS=NCCL pip install --global-option build_ext --global-option --debug -v -e .
```

This seems to have lead to two invocations of cmake:
1. `Running setup.py develop for horovod`, proceeding to do a debug build
2. `running build_ext`, proceeding to compile everything in _RelWithDebInfo_ mode

Then there was a weird error message:
```
    [100%] Built target horovod_cuda_kernels    
    make[1]: Leaving directory '/mnt/data/max/horovod/build/temp.linux-x86_64-3.7/RelWithDebInfo'
    /learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/cmake/data/bin/cmake -E cmake_progress_start /mnt/data/max/horovod/build/temp.linux-x86_64-3.7/RelWithDebInfo/CMakeFiles 0
    error: can't copy 'build/lib.linux-x86_64-3.7/horovod/tensorflow/mpi_lib.cpython-37m-x86_64-linux-gnu.so': doesn't exist or not a regular file
ERROR: Command errored out with exit status 1: /learndata4/maxDev/horovod-dev-venv/bin/python3 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/mnt/data/max/horovod/setup.py'"'"'; __file__='"'"'/mnt/data/max/horovod/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' build_ext --debug develop --no-deps Check the logs for full command output.
Exception information:
Traceback (most recent call last):
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 180, in _main
    status = self.run(options, args)
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/cli/req_command.py", line 205, in wrapper
    return func(self, options, args)
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 402, in run
    pycompile=options.compile,
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/req/__init__.py", line 85, in install_given_reqs
    pycompile=pycompile,
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/req/req_install.py", line 773, in install
    unpacked_source_directory=self.unpacked_source_directory,
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/operations/install/editable_legacy.py", line 46, in install_editable
    cwd=unpacked_source_directory,
  File "/learndata4/maxDev/horovod-dev-venv/lib/python3.7/site-packages/pip/_internal/utils/subprocess.py", line 244, in call_subprocess
    raise InstallationSubprocessError(proc.returncode, command_desc)
pip._internal.exceptions.InstallationSubprocessError: Command errored out with exit status 1: /learndata4/maxDev/horovod-dev-venv/bin/python3 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/mnt/data/max/horovod/setup.py'"'"'; __file__='"'"'/mnt/data/max/horovod/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' build_ext --debug develop --no-deps Check the logs for full command output.
```

Apparently it tried to copy the library for TensorFlow although I set `HOROVOD_WITHOUT_TENSORFLOW=1`.

I've attached a log file after running that a second time so all the gcc invocations are skipped over.
[build_log.txt](https://github.com/horovod/horovod/files/6938575/build_log.txt)

I don't really understand how options are supposed to be passed around between pip and setup.py... If we can't find a "canonical" way to configure a debug build for `pip -e`, we might just want to control that via a bespoke environment variable like `HOROVOD_DEBUG`.

Reply:
Thanks For trying it @maxhgerlach. I did a quick try in debug mode but did not let the build finish. Let me look into it.

Reply:
@maxhgerlach Try again with `HOROVOD_DEBUG=1`.

Reply:
Thanks @nvcastet, now it works like a charm for me.